File: Script_nested_dictionaries.jl
using Dates

function get_words_by_file(in_dir, n_files)
    cd(in_dir)
    filenames = filter(x -> endswith(lowercase(x), "txt"), readdir())[1:n_files]
    all_wds_by_file = Dict{String, Dict}()
    n_wds_corpus = 0
    for filename in filenames
        all_wds_by_file[filename] = Dict{String, Dict}()
        open(filename) do infile
            txt = read(infile, String)
            wds = [m.match for m in eachmatch(r"[-'’a-z]+"i, uppercase(txt))]
            n_wds_file = length(wds)
            n_wds_corpus += n_wds_file
            for i in 1:n_wds_file
                node = wds[i]
                if !haskey(all_wds_by_file[filename], node)
                    all_wds_by_file[filename][node] = Dict{String, Dict}("pre" => Dict{String, Int}(), "post" => Dict{String, Int}())
                end
                if i == 1
                    pre_wd = "<s>"
                else
                    pre_wd = wds[i-1]
                end
                if i == n_wds_file
                    post_wd = "<s>"
                else
                    post_wd = wds[i+1]
                end
                all_wds_by_file[filename][node]["pre"][pre_wd] = get(all_wds_by_file[filename][node]["pre"], pre_wd, 0) + 1
                all_wds_by_file[filename][node]["post"][post_wd] = get(all_wds_by_file[filename][node]["post"], post_wd, 0) + 1
            end  # next word
        end  # close connection to current file
    end  # next filename
    println("Julia: N words in corpus = ", n_wds_corpus)
    return all_wds_by_file

end  # end function

in_dir = "/path/to/dir"
open("/path/to/times.csv", "a") do outfile
    for i in collect(1000:1000:5000)
        println("Working on $i files...")
        for j in 1:10
            println("\tWorking on iteration $j")
            t1 = time()
            result = get_words_by_file(in_dir, i)
            duration = time() - t1
            println("\tJulia took ", duration, " seconds")
            println("\t", result["06vTDY97DIwqHxQSMMxZLq.txt"]["DAY"]) 
            write(outfile, "julia_v1.10.4,$i,$j,$duration\n")  
        end
    end  # next iteration    
end  # close connection to outfile

File: Script_nested_dictionaries.py
# Python 3 script to create a deeply nested dictionary
# Python v3.12.4

import os, re
from time import time

def get_words_by_file(in_dir: str, n_files: int) -> dict:
    os.chdir(in_dir)
    filenames = sorted([f for f in os.listdir() if f.lower().endswith("txt")])[:n_files]
    all_wds_by_file = {}
    n_wds_corpus = 0
    for f in filenames:
        all_wds_by_file[f] = {}
        with open(f) as infile:
            txt = infile.read().upper()
            wds = re.findall(r"[-'’a-z]+", txt, flags=re.I)
            n_wds_file = len(wds)
            n_wds_corpus += n_wds_file
            for i in range(n_wds_file):
                node = wds[i]
                if node not in all_wds_by_file[f].keys():
                    all_wds_by_file[f][node] = {"pre": {}, "post": {}}
                if i == 0:
                    pre_wd = "<s>"
                else:
                    pre_wd = wds[i-1]
                if i == len(wds) - 1:
                    post_wd = "<s>"
                else:
                    post_wd = wds[i+1]
                all_wds_by_file[f][node]["pre"][pre_wd] = all_wds_by_file[f][node]["pre"].get(pre_wd, 0) + 1
                all_wds_by_file[f][node]["post"][post_wd] = all_wds_by_file[f][node]["post"].get(post_wd, 0) + 1
    print("\tPython: N words in corpus = ", n_wds_corpus)
    return all_wds_by_file

if __name__ == "__main__":
    in_dir = "/path/to/dir"
    with open("/path/to/times.csv", "w") as outfile:
        outfile.write(f"lang,n_files,iter,sec\n")
        for i in range(1000, 6000, 1000):
            print(f"Working on {i} files...")
            for j in range(10):
                print("\tWorking on iteration", j+1)
                t0=time()
                result = get_words_by_file(in_dir, i)
                duration = time()-t0
                print("\tPython took", duration, "seconds")
                outfile.write(f"python_v3.12.4,{i},{j+1},{duration}\n")
                print("\t", result["06vTDY97DIwqHxQSMMxZLq.txt"]["DAY"])
